Build: fully resolve import paths in transpiled files

jsnajdr opened the pull request:
This PR implements the same thing as #73422, but as part of esbuild `onResolve` callback, without a separate post-processing step with Babel.

Local imports like `import from './proxy'` are resolved either to `'./proxy.js'` or `'./proxy/index.js'`, depending on which file/directory exists. That way the imports become compatible with Node.js ESM import rules, and also with browsers' native ESM import rules. They both require full file paths, without any shortcuts.

I originally wanted to eliminate local imports completely in #73516, but that proved too ambitious and risky.

This esbuild approach is very simple and flexible. If we want to, we can easily modify it to make all imports use the `.mjs` or `.cjs` extension. That will be useful for creating fully compliant dual CJS/ESM packages.

Before we merge this, I'd like to fine-tune the algorithm a bit. Make sure that special paths like `'.'` and `'..'` are handled correctly, that imports with an existing non-JS extension (like `block.json`) are not handled with this code, etc. But I didn't want to delay the PR further. Let's test and review.

jsnajdr committed:
Build: fully resolve import paths in transpiled files

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1013,6 +1032,33 @@ async function transpilePackage( packageName ) {
+					);
+					if ( isDir ) {
+						return {
+							path: args.path + '/index.js',
```
Should we use `path.join` ? Not sure about the Windows compatibility of this slash.

dmsnell replied:
yes, but also Node normalizes all paths so Windows compatibility isn’t an issue.

> On Windows, both the forward slash (/) and backward slash (\) are accepted as path segment separators; however, the path methods only add backward slashes (\).

https://nodejs.org/api/path.html#pathsep

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1013,6 +1032,33 @@ async function transpilePackage( packageName ) {
+					! args.path.endsWith( '.js' )
+				) {
+					const baseDir = path.dirname( args.importer );
+					for ( const ext of [ '.js', '.jsx', '.ts', '.tsx' ] ) {
```
Should we reuse `SOURCE_EXTENSIONS` ? Maybe what we have currently as `SOURCE_EXTENSIONS` should be a separate constant that uses the raw array of extension values.

aduth replied:
Actually, I wonder if we should be using some internal property of ESBuild like [`resolveExtensions`](https://esbuild.github.io/api/#resolve-extensions).

To that point, is it possible to tap into ESBuild's internal resolution behavior so we don't have to reimplement this?

jsnajdr replied:
I rewrote the plugin to use esbuild's internal `build.resolve` function. In the `onResolve` callback, ask esbuild to resolve the import (preventing recursion with `pluginData`) and then only modify the returned path. By replacing the source extensions (`.tsx` and others) with the target extension, which is either `.js` or `.mjs`.

That simplifies the plugin a lot. I no longer need to `stat` physical files, think about caching etc.

jsnajdr replied:
> Should we reuse `SOURCE_EXTENSIONS` ?

Probably yes, but there are complications: the globs want a bracket expansion syntax without dots, `{js,jsx}`, and for my code it's more convenient to have an array with dots, `[ '.js', '.jsx' ]`. Etc.

Let's consider this after everything else in this PR is in perfect order 🙂

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1013,6 +1032,33 @@ async function transpilePackage( packageName ) {
+				) {
+					const baseDir = path.dirname( args.importer );
+					for ( const ext of [ '.js', '.jsx', '.ts', '.tsx' ] ) {
+						const isFile = await pathIsFile(
```
Is there enough likelihood of reuse that it'd be worth considering to cache this result?

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1013,6 +1032,33 @@ async function transpilePackage( packageName ) {
+						}
+					}
+
+					const isDir = await pathIsDirectory(
```
I think we could be more efficient to call `stat` once for the path and reuse the `fs.Stats` result for checking if it's a file or directory. With this code, we'll call `stat` twice for directory imports.

jsnajdr committed:
Use esbuild resolve function, output .cjs extensions

jsnajdr committed:
Rename CJS util files to .cjs

jsnajdr committed:
Convert unbrotli/inflate files to ESM

jsnajdr committed:
Convert package.json files to modules

jsnajdr committed:
Added changelog entries

jsnajdr commented:
Some progress I made today:
- added `files` field to all `package.json` files so that the package content is exactly specified. Among other things, helps a lot with running `attw` on the packages.
- switched most packages to `"type": "module"` and renamed CJS `.js` files to `.cjs`. ESM files keep their `.js` extension.

There are still some wrinkles to iron out with `.d.ts` files, but the CI checks are green, build succeeds and the PR is mostly ready.

Fixes both #73362 and #73363.

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1013,6 +1018,47 @@ async function transpilePackage( packageName ) {
+					for ( const ext of exts ) {
+						if ( relativePath.endsWith( ext ) ) {
+							relativePath =
+								relativePath.slice( 0, -ext.length ) + newExt;
```
Feels like a job for [`path.basename`](https://nodejs.org/api/path.html#pathbasenamepath-suffix) with second argument.


```suggestion
							relativePath =
								path.basename( relativePath, ext ) + newExt;
```

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1013,6 +1018,47 @@ async function transpilePackage( packageName ) {
+								relativePath.slice( 0, -ext.length ) + newExt;
+							break;
+						}
+					}
```
We could probably do this with RegExp, maybe a bit more efficiently too. Not as future-compatible to if we wanted to actually consistently define extensions somewhere as an array (could probably construct from array using `new RegExp` constructor).


```suggestion
					// Replace extension: if path ends with one of the extensions, replace it
					const exts = [ '.js', '.jsx', '.ts', '.tsx' ];
					const newExt =
						build.initialOptions.format === 'cjs' ? '.cjs' : '.js';
					relativePath = relativePath.replace( /\.[jt]sx?$/, '' ) + newExt;
```

jsnajdr replied:
I implemented this suggestion, using `.replace` with a regexp. Originally we needed the `exts` array because we were looping through it, trying each extensions and checking with `stat` if the file exists. But now the esbuild resolver does this job for us.

jsnajdr committed:
is-shallow-equal: don't use default export

jsnajdr committed:
env: add missing exports

jsnajdr committed:
Fix imports from react-autosize-textarea

manzoorwanijk commented on packages/is-shallow-equal/src/index.ts:
```diff
@@ -29,3 +26,7 @@ export default function isShallowEqual( a: unknown, b: unknown ): boolean {
 
 	return a === b;
 }
+
+// `isShallowEqual` is exported also as a named export because esbuild cannot
+// expose the default export from the `window.wp.isShallowEqual` global.
+export { isShallowEqual, isShallowEqualObjects, isShallowEqualArrays };
```
We should do the same for all the packages mentioned in https://github.com/WordPress/gutenberg/issues/59087#issuecomment-2513784454

jsnajdr replied:
> We should do the same for all the packages mentioned in https://github.com/WordPress/gutenberg/issues/59087#issuecomment-2513784454

All the other packages should be OK, because they:
- either have a single default export and the `wpScriptDefaultExport` flag. Then `module.exports` value is directly the exported function.
- or multiple named exports and no default. Then `module.exports` is an object with a few named field, none of them `default`.

Both types of modules can be easily mapped between CJS and ESM without any problems.

The `is-shallow-equal` module is a combination of default and named exports. That's the only case that needs fixing.

Maybe I'm wrong, I'll need to look more closely what exactly is the problem with `api-fetch` as reported in #59087.

jsnajdr committed:
Mark type-only exports so that transpilation skips them

manzoorwanijk commented:
I see some progress via the attw CLI. 

```bash
"@wordpress/api-fetch"

node10: 🟢 
node16 (from CJS): 👺 Masquerading as ESM
🥴 Internal resolution error
node16 (from ESM): 🥴 Internal resolution error
bundler: 🟢 
```

I still strongly think that we can consider using [tsup](https://tsup.egoist.dev/) to create separate ESM and CJS bundles along with the corresponding types.

jsnajdr committed:
Include patches dir in node_modules cache key

jsnajdr commented on .github/setup-node/action.yml:
```diff
@@ -28,7 +28,7 @@ runs:
           with:
               path: '**/node_modules'
-              key: node_modules-${{ runner.os }}-${{ runner.arch }}-${{ steps.node-version.outputs.NODE_VERSION }}-${{ hashFiles('package-lock.json') }}
+              key: node_modules-${{ runner.os }}-${{ runner.arch }}-${{ steps.node-version.outputs.NODE_VERSION }}-${{ hashFiles('package-lock.json', 'patches/**') }}
```
I had to do this when I modified the `patches/react-autosize-textarea+7.1.0.patch` file. It modifies the cache key for the `node_modules` cache so that it includes also a hash of all the `patches/**` files. These patches represent a modification of the `node_modules` folder.

Without it, the `patch-package` tool in `postinstall` step was failing. A copy of `node_modules` was retrieved that already had the patches applied. Then `patch-package` is run on it again. If the patches are already applied, the tool detects that and finishes successfully. But if an old patch is applied and we're trying to apply a new one, then they conflict and the step fail, breaking the entire CI pipeline.

jsnajdr commented on packages/admin-ui/package.json:
```diff
@@ -22,13 +22,22 @@
+		"*.md"
+	],
+	"type": "module",
+	"main": "build/index.cjs",
```
This is a typical change I'm doing to most `package.json` files:
- explicitly list `files` so that i) `npm pack` can reliably produce a package tarball, including .gitignored folders like `build-module`; ii) we don't ship unneeded files like `tsconfig.tsbuildinfo`.
- add `type: module`
- rename the CJS entrypoint to `.cjs`

jsnajdr commented on packages/api-fetch/src/index.ts:
```diff
@@ -205,4 +205,4 @@ apiFetch.mediaUploadMiddleware = mediaUploadMiddleware;
 
 export default apiFetch;
-export * from './types';
+export type * from './types';
```
This helps esbuild completely eliminate the `./types` import when transpiling. Without this, it imports the `./types` module which is competely empty, and reexports "everything", which turns out to be "nothing". Harmless but useless.

aduth replied:
I see a couple other `export * from './types';`. Should those be updated?

https://github.com/WordPress/gutenberg/blob/94d3842d0812e543dc25bdf7b4f314d6071ed997/packages/shortcode/src/index.js#L6

https://github.com/WordPress/gutenberg/blob/94d3842d0812e543dc25bdf7b4f314d6071ed997/packages/wordcount/src/index.ts#L140

jsnajdr commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -4,7 +4,7 @@
 import { useRegistry } from '@wordpress/data';
 import deprecated from '@wordpress/deprecated';
-import isShallowEqual from '@wordpress/is-shallow-equal';
+import { isShallowEqual } from '@wordpress/is-shallow-equal';
```
Adding `isShallowEqual` as a named export in the package, in addition to the existing default export, and using the named export consistently, this all helps a lot with CJS/ESM reliability.

When an ESM module `imports` a CJS module, it works, because both Node.js and bundlers have interoperability, but `default` exports are interpreted differently in some cases. Typically when a module has _both_ default and named exports. That's why we should avoid this "default+named" layout.

jsnajdr commented on packages/block-library/src/utils/is-block-metadata-experimental.js:
```diff
@@ -2,18 +2,13 @@
 		metadata.__experimental !== false
 	);
-};
+}
```
With `type: module`, this CJS module now must be converted to ESM.

In the `babel-plugin.cjs` file I'm inlining the code, because CJS modules can't import ESM modules.

We can eventually fix this by converting the Babel plugin, and most other CJS modules, to ESM. So that our codebase is fully ESM and has zero CJS files.

anomiex replied:
> In the `babel-plugin.cjs` file I'm inlining the code, because CJS modules can't import ESM modules.

FYI, Node 20.19+, 22.12+, and 24+ allow importing ESM modules as long as they don't have any top-level `await` (directly or in any imports).

jsnajdr replied:
Good to know 🙂 Gutenberg is still on v20, and with my local v20.16.0 I was getting:
```
Error [ERR_REQUIRE_ESM]: [BABEL]: require() of ES Module src/utils/is-block-metadata-experimental.js from babel-plugin.cjs not supported.
Instead change the require of is-block-metadata-experimental.js in babel-plugin.cjs to a dynamic import() which is available in all CommonJS modules.
```

anomiex replied:
Yeah, 20.19 is the first v20 version that has it not behind a feature flag. 🙂

jsnajdr commented on packages/block-library/babel-plugin.cjs:
```diff
@@ -3,11 +3,6 @@
-/**
- * Internal dependencies
- */
-const isBlockMetadataExperimental = require( './src/utils/is-block-metadata-experimental' );
```
This `require` now fails because Node.js doesn't support requiring ESM from CJS:
```
Error [ERR_REQUIRE_ESM]: [BABEL]: require() of ES Module src/utils/is-block-metadata-experimental.js from babel-plugin.cjs not supported.
Instead change the require of is-block-metadata-experimental.js in babel-plugin.cjs to a dynamic import() which is available in all CommonJS modules.
```

I decided to inline the code because it's very simple.

jsnajdr commented on packages/global-styles-ui/src/font-library/lib/inflate.js:
```diff
@@ -30,25 +30,8 @@
-	}
-} )( function () {
+/* pako 1.0.10 nodeca/pako */
+export default ( function () {
```
This converts the 3rd party vendored module from UMD to ESM. Turns out to be a very simple change.

jsnajdr commented on packages/scripts/scripts/test-playwright.js:
```diff
@@ -40,7 +39,7 @@ let loadConfig = null;
 try {
 	// First, try to load the package installed from among the optional peerDependencies.
-	loadConfig = require( '@wordpress/env/lib/config/load-config' );
+	loadConfig = require( '@wordpress/env/lib/config' ).loadConfig;
```
This fixes a bug introduced by the build v2 migration in #72251. The `env` package didn't have the `lib/config` exports that are used here, and the import was always failing.

jsnajdr commented on packages/scripts/scripts/test-playwright.js:
```diff
@@ -14,7 +14,6 @@ process.on( 'unhandledRejection', ( err ) => {
  */
 const { resolve } = require( 'node:path' );
 const { sync: spawn } = require( 'cross-spawn' );
-const path = require( 'path' );
```
Here the Node.js `path` module was imported twice, I'm fixing that.

jsnajdr commented on patches/react-autosize-textarea+7.1.0.patch:
```diff
@@ -11,3 +11,13 @@ index ad6ff3f..e0a6d51 100644
+-exports.__esModule = true;
+ var TextareaAutosize_1 = require("./TextareaAutosize");
+-exports["default"] = TextareaAutosize_1.TextareaAutosize;
++module.exports = TextareaAutosize_1.TextareaAutosize;
```
This patches the `react-autosize-textarea` package so that it doesn't use the incompatible "object with default export and `__esModule`" pattern, but just assigns the default export directly to `module.exports`. That's what Node.js ESM/CJS interop understands.

jsnajdr committed:
Replace extensions with a regexp

jsnajdr commented:
> I had seen a library (can't remember which one) using a script to copy .d.ts → .d.mts or .d.cts after the build, since the content is often identical.

Yes, I tried that locally and it worked, it fixes the "masquerading as ESM" issue. But there's still the issue that `.d.ts` files have imports like `./foo` that are not expanded to the full `./foo/index.d.ts` paths. The most straightforward fix is probably that we migrate our codebase to fully specified imports. Then the source files will be already compliant, and build steps have nothing to fix.

jsnajdr commented on packages/wp-build/src/build.mjs:
```diff
@@ -1033,6 +1038,40 @@ async function transpilePackage( packageName ) {
 					return null;
 				}
 
+				// Fully resolve local dependencies without file extension
+				// and replace the extension with the target extension.
+				if ( args.path.startsWith( '.' ) ) {
+					const resolved = await build.resolve( args.path, {
+						namespace: args.namespace,
+						importer: args.importer,
+						kind: args.kind,
+						resolveDir: args.resolveDir,
+						pluginData: { __fromExternalize: true },
+					} );
+
+					if ( resolved.errors.length > 0 ) {
+						return resolved;
+					}
+
+					// Relativize path: make it relative to resolveDir with leading ./
+					let relativePath = normalizePath(
+						path.relative( args.resolveDir, resolved.path )
+					);
+					if ( ! relativePath.startsWith( '.' ) ) {
+						relativePath = './' + relativePath;
+					}
+
+					// Replace extension: make sure that file extension is always `.js` or `.cjs`.
+					const newExt =
+						build.initialOptions.format === 'cjs' ? '.cjs' : '.js';
+					relativePath = relativePath.replace( /\.[jt]sx?$/, newExt );
+
+					return {
+						path: relativePath,
+						external: true,
+					};
+				}
```
This works as expected. If there is an import like `./file.json`, we still want to process it with this hook. Resolve to full file path and mark as external. The only step we _don't_ want to do is the extension replacement. Because the `.json` extension is already fine, we don't need to change it. We only want do change `.tsx` to `.js` for example.

When there is an import like `./file` and there is a matching `./file.json` file, the esbuild resolver will find it. Esbuild has a [resolveExtensions](https://esbuild.github.io/api/#resolve-extensions) whose default value includes `.json`. So, the `./file` import is going to be expanded into `./file.json` and externalized. That's the useful work the hook has done on it, even if it didn't replace extensions.

jsnajdr commented on packages/wp-build/src/build.mjs:
```diff
@@ -1033,6 +1038,40 @@ async function transpilePackage( packageName ) {
 
+				// Fully resolve local dependencies without file extension
+				// and replace the extension with the target extension.
+				if ( args.path.startsWith( '.' ) ) {
```
`import '.'` and `import '..'` are relative paths like any other, they will _correctly_ enter this resolution logic. They will be resolved into something like `../index.js`, the `import` statement will be modified to import the full path, and will be marked as external.

jsnajdr commented on packages/core-abilities/package.json:
```diff
@@ -24,19 +24,21 @@
 		"build",
 		"build-module",
 		"build-types",
 		"src",
```
These are nice finds, I fixed them 🙂

aduth commented on packages/a11y/package.json:
```diff
@@ -24,13 +24,21 @@
 	},
-	"main": "build/index.js",
+	"files": [
+		"src",
```
In the future, we might want to revisit if we can exclude `src`. Since `exports` is an enforced public API, we should be able to mostly limit to what's referenced through those entrypoints.

aduth commented on packages/env/package.json:
```diff
@@ -35,6 +35,7 @@
 	"exports": {
 		".": "./lib/env.js",
 		"./lib/cli.js": "./lib/cli.js",
+		"./lib/config": "./lib/config/index.js",
```
What is this for? Just noting the inconsistency with the file extension included for `cli.js`. Ideally we could call this like `/cli` and `/config` if needed, but I guess that ship has sailed.

jsnajdr replied:
This fixes a bug described in https://github.com/WordPress/gutenberg/pull/73822#discussion_r2630053313. That functionality didn't work. Trying to load a custom `.wp-env` config always failed even if the config file existed, because loading the `load-config` module failed.

This could be improved in a little follow-up PR. The `lib/cli` export is not needed at all, because the `bin` script can import it anyway. And the `lib/config` export can be merged into the root `.` one.

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1033,6 +1038,40 @@ async function transpilePackage( packageName ) {
 
+				// Fully resolve local dependencies without file extension
+				// and replace the extension with the target extension.
+				if ( args.path.startsWith( '.' ) ) {
```
I was curious if it'd be an optimization to use index access to check the first character. Interestingly they're about the same!

```
startsWith x 298,278,099 ops/sec ±5.39% (76 runs sampled)
index x 291,990,356 ops/sec ±5.14% (76 runs sampled)
```

```ts
import Benchmark from "benchmark";

const suite = new Benchmark.Suite();

suite.add("startsWith", () => {
  "./foo".startsWith(".");
});

suite.add("index", () => {
  "./foo"[0] === ".";
});

suite
  .run({ async: true })
  .on("cycle", (event) => {
    console.log(String(event.target));
  })
  .on("complete", () => {
    console.log("Fastest is " + suite.filter("fastest").map("name"));
  });
```

aduth commented on packages/wp-build/src/build.mjs:
```diff
@@ -1033,6 +1038,40 @@ async function transpilePackage( packageName ) {
 					return null;
 				}
 
+				// Fully resolve local dependencies without file extension
+				// and replace the extension with the target extension.
+				if ( args.path.startsWith( '.' ) ) {
+					const resolved = await build.resolve( args.path, {
+						namespace: args.namespace,
+						importer: args.importer,
+						kind: args.kind,
+						resolveDir: args.resolveDir,
+						pluginData: { __fromExternalize: true },
+					} );
```
1. Do we need to list the properties or can we just forward it blindly?
2. Do we need to carry forward any `pluginData` that might already be in `args.pluginData` ?


```suggestion
					const resolved = await build.resolve( args.path, {
						...args,
						pluginData: { ...args.pluginData, __fromExternalize: true },
					} );
```

jsnajdr replied:
I implemented carrying over the `pluginData`, but forwarding all the options is IMO not so good. These are two different types: `OnResolveArgs` and `ResolveOptions`. Mostly the same, but not 100%. That's why I'm enumerating them.

aduth reviewed (approved):
These changes make sense and appear to work well in my testing. I left a few minor comments, but overall LGTM 👍 I noticed the transpilation is ~30% slower than `trunk` (probably from the `build.resolve`), but feels necessary for correctness and better than the alternatives we considered.

youknowriad commented:
> I noticed the transpilation is ~30% slower than trunk 

I'm starting to think that we should have two "build" modes. 

 1- One for day to day work, target Gutenberg plugin and/or WordPress.
 2- Full build with types and all.

For the former we can assume "less TS" and maybe less of these "correctness" issues.
For the latter, it's the full spectrum.

I see the former maybe being useful for `npm run dev` or for building Gutenberg within WordPress Core... where you're not really shipping packages, you don't really needs .d.ts files either...

jsnajdr committed:
onResolve: forward pluginData and with

jsnajdr commented:
> I see a couple other `export * from './types';`

I updated the `wordcount` one. The `shortcode` file has an `.js` extension, can't use the `type` syntax there. Let's do it when it's converted.

jsnajdr committed:
Fix unrelated e2e test

jsnajdr commented:
> `jest-console` having no default export is broken

Yes, the CJS version of `jest-console` is broken, importing internal CJS modules with default exports (`import supportedMatchers from './supported-matchers';`) doesn't work. The `__toESM` helper produces a double default: `{ default: { default: obj } }`. I'm looking into this.

It's very likely that many other CJS modules are broken the same way.

jsnajdr commented:
We're running into this issue: https://github.com/evanw/esbuild/issues/2623

When a file like `jest-console/src/index.js` imports a default export from a local submodule:
```js
import supportedMatchers from './supported-matchers';
```
Then the transpiled CJS files' `require` is broken. The object resulting from `require('supported-matchers.cjs')` is wrong. It has the `{ default: { default: matchers }` shape instead of `{ default: matchers }`.

The advice in the PR is to never use default exports. That's not great. I'm trying to figure out what else we can do.
